src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -89,162 +99,217 @@ public class IndexDescriptor
-     * If context is null, the component is assumed to be a per-sstable component.
-     */
-    private static boolean componentExistsOnDisk(Version version, Descriptor descriptor, IndexComponent component, IndexContext context)
+    public IndexComponents.ForWrite newPerIndexComponentsForWrite(IndexContext context)
     {
-        var file = fileFor(descriptor, version, component, context);
-        return file.exists();
+        maybePopulateComponents(this, context);
+        return newComponentsForWrite(context);
     }
 
-    public File fileFor(IndexComponent component)
+    private IndexComponents.ForWrite newComponentsForWrite(@Nullable IndexContext context)
     {
-        var ac = new AttachedIndexComponent(component, null);
-        return fileMap.computeIfAbsent(ac, __ -> createFile(component, null));
+        var currentComponents = groups.get(context);
+        // If we're "bumping" the version compared to the existing group, then we can use generation 0. Otherwise, we
+        // have to bump the generation, unless we're using immutable components, in which case we always use generation 0.
+        // Unless we don't use immutable components, in which case we always use generation 0.
+        Version newVersion = Version.latest();
+        if (newVersion.useImmutableComponentFiles())
+        {
+            int candidateGeneration = currentComponents.version().equals(newVersion)
+                                      ? currentComponents.generation() + 1
+                                      : 0;
+            // Usually, we'll just use `candidateGeneration`, but we want to avoid overriding existing file (it's
+            // theoretically possible that the next generation was created at some other point, but then corrupted,
+            // and so we falled back on the previous generation but some of those file for the next generation still
+            // exists). So we check repeatedly increment the generation until we find one for which no files exist.
+            return createFirstGenerationAvailableComponents(context, newVersion, candidateGeneration);
+        }
+        else
+        {
+            return new IndexComponentsImpl(context, newVersion, 0);
+        }
     }
 
-    public File fileFor(IndexComponent component, IndexContext context)
+    private IndexComponentsImpl createFirstGenerationAvailableComponents(@Nullable IndexContext context, Version version, int startGeneration)
     {
-        return fileMap.computeIfAbsent(new AttachedIndexComponent(component, context),
-                                                     p -> createFile(component, context));
+        int generationToTest = startGeneration;
+        while (true)
+        {
+            IndexComponentsImpl candidate = new IndexComponentsImpl(context, version, generationToTest);
+            if (candidate.expectedComponentsForVersion().stream().noneMatch(candidate::componentExistsOnDisk))
```

Comment:
would be nice to have some logs here about generation conflict..

Reply:
Added some.